python/lsst/ap/pipe/ap_pipe.py

```diff
@@ -21,356 +21,327 @@
+            "r": "r",
+            "i": "i",
+            "z": "z",
+            "y": "y",
+            "VR": "g"}
+
+        # TODO: single-template support now done by retargeting self.differencer.getTemplate
+        # Document how to do this in DM-13164
+        self.differencer.detection.thresholdValue = 5.0
+        self.differencer.doDecorrelation = True
+        self.differencer.coaddName = "deep"  # TODO: generalize in DM-12315
+        self.differencer.getTemplate.warpType = "psfMatched"
+        self.differencer.doSelectSources = False
+
+        self.associator.level1_db.retarget(AssociationDBSqliteTask)
+
+    def validate(self):
+        pexConfig.Config.validate(self)
+
+
+class ApPipeTask(pipeBase.CmdLineTask):
+    """`CmdLineTask` representing the entire AP pipeline.
+
+    ``ApPipeTask`` processes raw DECam images from basic processing through
+    source association. Other observatories will be supported in the future.
+
+    A tutorial for using ``ApPipeTask`` is available in
+    [DMTN-039](http://dmtn-039.lsst.io).
+
+    ``ApPipeTask`` can be run from the command line, but it can also be called
+    from other pipeline code. It provides public methods for executing each
+    major step of the pipeline by itself.
+    """
+
+    ConfigClass = ApPipeConfig
+    RunnerClass = ApPipeTaskRunner
+    _DefaultName = "apPipe"
+
+    # TODO: dbFile is a workaround for DM-11767
+    def __init__(self, butler, dbFile, config=None, *args, **kwargs):
+        """Set up the AP pipeline.
+
+        Parameters
+        ----------
+        butler : `lsst.daf.persistence.Butler`
+            A Butler providing access to the science, calibration, and (unless
+            ``config.differencer.getTemplate`` is overridden) template data to
+            be processed. Its output repository must be both readable
+            and writable.
+        dbFile : `str`
+            The filename where the source association database lives. Will be
+            created if it does not yet exist.
+        config : `ApPipeTask.ConfigClass`, optional
+            A configuration for this task. Defaults to ApPipeTask.ConfigClass().
+        """
+        # TODO: hacky workaround for DM-13602
+        modConfig = ApPipeTask._copyConfig(config) if config is not None else ApPipeTask.ConfigClass()
+        modConfig.associator.level1_db.db_name = dbFile
+        modConfig.freeze()
+        pipeBase.CmdLineTask.__init__(self, *args, config=modConfig, **kwargs)
+
+        self.makeSubtask("ccdProcessor", butler=butler)
+        self.makeSubtask("differencer", butler=butler)
+        # Must be called before AssociationTask.__init__
+        _setupDatabase(self.config.associator.level1_db)
+        self.makeSubtask("associator")
+
+    # TODO: hack for modifying frozen configs; delete once DM-13602 resolved
+    @staticmethod
+    def _copyConfig(config):
+        configClass = type(config)
+        contents = {key: value for (key, value) in config.items()}
```

Comment:
It turns out I did try it. I get a `FieldValidationError` from the new config's constructor because `toDict()` is recursive and the constructor expects a `ProcessCcdConfig`, not a `dict`. That's why I wrote it this way -- to force only the top level to be converted.